File: training/normalize.py
"""Python port of src/features/normalize.ts.

The TypeScript is the source of truth: it is what runs at inference time. This
exists so training sees exactly the same feature vectors the browser will, and
``test_parity.py`` checks that claim against fixtures.

If the two ever disagree, fix this file, not the TypeScript.
"""

from __future__ import annotations

import numpy as np

WRIST = 0
THUMB_TIP = 4
INDEX_MCP = 5
INDEX_TIP = 8
MIDDLE_MCP = 9
MIDDLE_TIP = 12
RING_TIP = 16
PINKY_MCP = 17
PINKY_TIP = 20

FINGER_CHAINS = {
    "thumb": (1, 2, 3, 4),
    "index": (5, 6, 7, 8),
    "middle": (9, 10, 11, 12),
    "ring": (13, 14, 15, 16),
    "pinky": (17, 18, 19, 20),
}

EPS = 1e-6


def hand_span(landmarks: np.ndarray) -> float:
    """Wrist to middle-MCP: the most articulation-stable segment on the hand.

    A closed fist and a flat palm give nearly the same value, which is exactly
    what a scale reference needs.
    """
    primary = float(np.linalg.norm(landmarks[MIDDLE_MCP] - landmarks[WRIST]))
    if primary > EPS:
        return primary
    across = float(np.linalg.norm(landmarks[INDEX_MCP] - landmarks[PINKY_MCP]))
    return across if across > EPS else 1.0


def normalize_hand(
    landmarks: np.ndarray,
    handedness: str = "Right",
    aspect: float = 1.0,
    mirror_left: bool = True,
    canonical_rotation: bool = True,
) -> np.ndarray:
    """Normalize one hand to canonical pose-invariant space.

    Steps, in the same order as normalize.ts:
      aspect-correct -> mirror -> translate -> scale -> rotate.

    Args:
        landmarks: (21, 3) array of image-normalized coordinates.

    Returns:
        (21, 3) array. Wrist at the origin, unit hand span, canonical roll.
    """
    pts = np.asarray(landmarks, dtype=np.float64).copy()
    if pts.shape != (21, 3):
        raise ValueError(f"expected (21, 3) landmarks, got {pts.shape}")

    # 1. aspect correction
    pts[:, 0] *= aspect
    pts[:, 2] *= aspect

    # 2. mirror left hands into right-hand space
    if mirror_left and handedness == "Left":
        pts[:, 0] *= -1

    # 3. wrist to origin
    pts -= pts[WRIST]

    # 4. unit hand span
    pts /= hand_span(pts)

    # 5. canonical roll: wrist -> middle-MCP onto +y
    if canonical_rotation:
        ref = pts[MIDDLE_MCP]
        length = float(np.hypot(ref[0], ref[1]))
        if length > EPS:
            cos = ref[1] / length
            sin = ref[0] / length
            x = pts[:, 0] * cos - pts[:, 1] * sin
            y = pts[:, 0] * sin + pts[:, 1] * cos
            pts[:, 0] = x
            pts[:, 1] = y

    return pts


def feature_vector(normalized: np.ndarray) -> np.ndarray:
    """Flatten to the 63-float vector the classifiers consume."""
    return normalized.reshape(-1).astype(np.float32)


def hand_features(
    landmarks: np.ndarray, handedness: str = "Right", aspect: float = 1.0
) -> np.ndarray:
    """Landmarks straight to a 63-float feature vector."""
    return feature_vector(normalize_hand(landmarks, handedness, aspect))


# --- sign-level windowing (mirrors src/features/window.ts) -------------------

WINDOW_FRAMES = 64
PER_FRAME_DIM = 63 + 63 + 6 + 2
LEFT_SHOULDER = 11
RIGHT_SHOULDER = 12


def body_frame(pose: np.ndarray | None, aspect: float = 1.0):
    """Shoulder-derived reference frame: origin between shoulders, scale = width."""
    if pose is None or len(pose) <= RIGHT_SHOULDER:
        return None
    left = pose[LEFT_SHOULDER]
    right = pose[RIGHT_SHOULDER]
    width = float(np.hypot((left[0] - right[0]) * aspect, left[1] - right[1]))
    if width < 1e-4:
        return None
    return ((left[0] + right[0]) / 2, (left[1] + right[1]) / 2, width)


def frame_features(
    dominant: np.ndarray | None,
    other: np.ndarray | None,
    pose: np.ndarray | None,
    aspect: float = 1.0,
    dominant_handedness: str = "Right",
    other_handedness: str = "Left",
) -> np.ndarray:
    """One frame of the sign feature vector. Layout matches window.ts exactly."""
    out = np.zeros(PER_FRAME_DIM, dtype=np.float32)
    frame = body_frame(pose, aspect)

    offset = 0
    for hand, handedness in ((dominant, dominant_handedness), (other, other_handedness)):
        if hand is not None:
            out[offset : offset + 63] = hand_features(hand, handedness, aspect)
        offset += 63

    for hand in (dominant, other):
        if hand is not None:
            wrist = hand[WRIST]
            if frame is None:
                out[offset : offset + 3] = [wrist[0] - 0.5, wrist[1] - 0.5, wrist[2]]
            else:
                ox, oy, scale = frame
                out[offset : offset + 3] = [
                    (wrist[0] - ox) / scale,
                    (wrist[1] - oy) / scale,
                    wrist[2] / scale,
                ]
        offset += 3

    out[offset] = 1.0 if dominant is not None else 0.0
    out[offset + 1] = 1.0 if other is not None else 0.0
    return out


def resample_window(frames: np.ndarray, target: int = WINDOW_FRAMES) -> np.ndarray:
    """Resample (T, D) to (target, D) with linear interpolation.

    Resampling rather than padding is what makes a sign performed quickly and
    the same sign performed slowly produce the same feature vector. Speed is a
    signer trait, not a lexical one.
    """
    frames = np.asarray(frames, dtype=np.float32)
    if frames.ndim != 2:
        raise ValueError(f"expected (T, D), got {frames.shape}")
    if len(frames) == 0:
        return np.zeros((target, PER_FRAME_DIM), dtype=np.float32)
    if len(frames) == 1:
        return np.repeat(frames, target, axis=0)

    source = np.linspace(0.0, 1.0, len(frames))
    wanted = np.linspace(0.0, 1.0, target)
    out = np.empty((target, frames.shape[1]), dtype=np.float32)
    for d in range(frames.shape[1]):
        out[:, d] = np.interp(wanted, source, frames[:, d])
    return out

File: src/features/normalize.ts
/**
 * Landmark normalization — pure math, no DOM, no state.
 *
 * Every downstream classifier consumes the output of this module, so a change
 * here invalidates every trained model. The unit tests in tests/normalize.test.ts
 * pin the exact behaviour; treat them as the specification.
 *
 * Pipeline, in order:
 *   1. aspect-correct    image-normalized coords are stretched by the frame's
 *                        aspect ratio; undo that so distances are isotropic
 *   2. mirror            left hands are reflected into right-hand space so one
 *                        model serves both dominant hands
 *   3. translate         wrist (landmark 0) becomes the origin
 *   4. scale             divide by hand span so distance-to-camera drops out
 *   5. rotate            wrist -> middle-MCP points along +y, canonical roll
 */
import type { HandFrame, Point3 } from '@/vision/types';
import { HAND_LANDMARK } from '@/vision/types';

export interface NormalizeOptions {
  /** Frame aspect ratio (width / height). 1 disables aspect correction. */
  aspect?: number;
  /** Reflect left hands into right-hand space. Default true. */
  mirrorLeft?: boolean;
  /** Rotate so the hand's long axis is canonical. Default true. */
  canonicalRotation?: boolean;
}

const EPS = 1e-6;

/** Euclidean distance in 3-space. */
export function dist(a: Point3, b: Point3): number {
  const dx = a.x - b.x;
  const dy = a.y - b.y;
  const dz = a.z - b.z;
  return Math.sqrt(dx * dx + dy * dy + dz * dz);
}

/** Distance ignoring depth — MediaPipe's z is far noisier than x/y. */
export function dist2d(a: Point3, b: Point3): number {
  const dx = a.x - b.x;
  const dy = a.y - b.y;
  return Math.sqrt(dx * dx + dy * dy);
}

/**
 * Reference length used to scale a hand to unit size.
 *
 * Wrist -> middle-MCP is the most stable segment on the hand: it barely changes
 * with finger articulation, unlike anything involving a fingertip. A closed fist
 * and a flat palm produce nearly the same value, which is exactly what we want.
 */
export function handSpan(landmarks: Point3[]): number {
  const wrist = landmarks[HAND_LANDMARK.WRIST];
  const midMcp = landmarks[HAND_LANDMARK.MIDDLE_MCP];
  const primary = dist(wrist, midMcp);
  if (primary > EPS) return primary;
  // Degenerate hand (all points coincident): fall back to knuckle width.
  const across = dist(landmarks[HAND_LANDMARK.INDEX_MCP], landmarks[HAND_LANDMARK.PINKY_MCP]);
  return across > EPS ? across : 1;
}

/**
 * Normalize one hand to a canonical pose-invariant frame.
 * Returns a fresh array; the input is never mutated.
 */
export function normalizeHand(
  landmarks: Point3[],
  handedness: 'Left' | 'Right',
  opts: NormalizeOptions = {},
): Point3[] {
  const { aspect = 1, mirrorLeft = true, canonicalRotation = true } = opts;

  // 1. aspect correction — x spans `aspect` times the physical width of y.
  const pts: Point3[] = landmarks.map((p) => ({ x: p.x * aspect, y: p.y, z: p.z * aspect }));

  // 2. mirror left hands into right-hand space.
  if (mirrorLeft && handedness === 'Left') {
    for (const p of pts) p.x = -p.x;
  }

  // 3. translate wrist to origin.
  const wrist = { ...pts[HAND_LANDMARK.WRIST] };
  for (const p of pts) {
    p.x -= wrist.x;
    p.y -= wrist.y;
    p.z -= wrist.z;
  }

  // 4. scale to unit hand span.
  const span = handSpan(pts);
  const inv = 1 / span;
  for (const p of pts) {
    p.x *= inv;
    p.y *= inv;
    p.z *= inv;
  }

  // 5. rotate about z so wrist -> middle-MCP points along +y.
  if (canonicalRotation) {
    const ref = pts[HAND_LANDMARK.MIDDLE_MCP];
    const len = Math.hypot(ref.x, ref.y);
    if (len > EPS) {
      // Rotation that maps (ref.x, ref.y) onto (0, len).
      const cos = ref.y / len;
      const sin = ref.x / len;
      for (const p of pts) {
        const x = p.x * cos - p.y * sin;
        const y = p.x * sin + p.y * cos;
        p.x = x;
        p.y = y;
      }
    }
  }

  return pts;
}

/** Flatten normalized landmarks to the 63-float vector the classifiers expect. */
export function toFeatureVector(normalized: Point3[]): Float32Array {
  const out = new Float32Array(normalized.length * 3);
  for (let i = 0; i < normalized.length; i++) {
    out[i * 3] = normalized[i].x;
    out[i * 3 + 1] = normalized[i].y;
    out[i * 3 + 2] = normalized[i].z;
  }
  return out;
}

/** Convenience: HandFrame -> 63-float vector in one step. */
export function handFeatureVector(hand: HandFrame, opts: NormalizeOptions = {}): Float32Array {
  return toFeatureVector(normalizeHand(hand.landmarks, hand.handedness, opts));
}

/**
 * Where the hand sits in the frame, in raw image coordinates (0..1).
 * Used by the auto-space heuristic, which cares about absolute position and so
 * must not use the translated/scaled landmarks.
 */
export function handCentroid(landmarks: Point3[]): { x: number; y: number } {
  let x = 0;
  let y = 0;
  for (const p of landmarks) {
    x += p.x;
    y += p.y;
  }
  return { x: x / landmarks.length, y: y / landmarks.length };
}

/** Cosine similarity between two equal-length vectors. */
export function cosineSimilarity(a: Float32Array, b: Float32Array): number {
  let dot = 0;
  let na = 0;
  let nb = 0;
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) {
    dot += a[i] * b[i];
    na += a[i] * a[i];
    nb += b[i] * b[i];
  }
  const denom = Math.sqrt(na) * Math.sqrt(nb);
  return denom > EPS ? dot / denom : 0;
}

/** Squared L2 distance — cheaper than the real distance for nearest-centroid. */
export function squaredDistance(a: Float32Array, b: Float32Array): number {
  let sum = 0;
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) {
    const d = a[i] - b[i];
    sum += d * d;
  }
  return sum;
}

/** Elementwise mean of equal-length vectors. Returns null for an empty set. */
export function meanVector(vectors: Float32Array[]): Float32Array | null {
  if (vectors.length === 0) return null;
  const out = new Float32Array(vectors[0].length);
  for (const v of vectors) {
    for (let i = 0; i < out.length; i++) out[i] += v[i];
  }
  for (let i = 0; i < out.length; i++) out[i] /= vectors.length;
  return out;
}
